header/p2p/session.go

```diff
@@ -0,0 +1,224 @@
+package p2p
+
+import (
+	"context"
+	"errors"
+	"io"
+	"sort"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/protocol"
+
+	"github.com/celestiaorg/go-libp2p-messenger/serde"
+
+	"github.com/celestiaorg/celestia-node/header"
+	p2p_pb "github.com/celestiaorg/celestia-node/header/p2p/pb"
+)
+
+// TODO(@vgonkivs): make it configurable
+var (
+	// headersPerPeer is a maximum amount of headers that will be requested per peer
+	headersPerPeer uint64 = 64
+)
+
+// session aims to divide a range of headers
+// into several smaller requests among different peers.
+type session struct {
+	ctx        context.Context
+	cancel     context.CancelFunc
+	host       host.Host
+	protocolID protocol.ID
+	// peerTracker contains discovered peers with records that describes their activity.
+	queue *peerQueue
+
+	reqCh chan *p2p_pb.ExtendedHeaderRequest
+	errCh chan error
+}
+
+func newSession(ctx context.Context, h host.Host, peerTracker []*peerStat, protocolID protocol.ID) *session {
+	ctx, cancel := context.WithTimeout(ctx, time.Minute*2)
+	return &session{
+		ctx:        ctx,
+		cancel:     cancel,
+		protocolID: protocolID,
+		host:       h,
+		queue:      newPeerQueue(peerTracker),
+		errCh:      make(chan error),
+	}
+}
+
+// doRequest chooses the best peer to fetch headers and sends a request in range of available maxRetryAttempts
+func (s *session) doRequest(
+	ctx context.Context,
+	stat *peerStat,
+	req *p2p_pb.ExtendedHeaderRequest,
+	headers chan []*header.ExtendedHeader,
+) {
+	select {
+	case <-ctx.Done():
+		return
+	case <-s.ctx.Done():
+		return
+	default:
+	}
+	r, size, duration, err := s.requestHeaders(ctx, stat.peerID, req)
+	if err != nil {
+		if err == context.Canceled {
+			return
+		}
+		log.Errorw("requesting headers from peer failed."+
+			"Retrying the request from different peer", "failed peer", stat.peerID, "err", err)
+		s.reqCh <- req
```

Comment:
🚓
Sorry for being annoying, but we need to wrap this with select as well, so that we never leak routines if context is canceled.